tests/unit/test_state.py

```diff
@@ -0,0 +1,112 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Jenkins-k8s state module tests."""
+
+from unittest.mock import MagicMock
+
+import pytest
+
+import state
+
+
+@pytest.mark.parametrize(
+    "time_range",
+    [
+        pytest.param("-1-3", id="negative time range"),
+        pytest.param("00-24", id="Out of 24H scale"),
+        pytest.param("23", id="Not a range"),
+        pytest.param("23-23", id="Same time given as range"),
+        pytest.param("00:01-00:02", id="Hour:minute format"),
+        pytest.param("3PM-4PM", id="Non 24H format"),
+        pytest.param("01--02", id="Invalid input (extra dash)"),
+        pytest.param("2023‐06‐20T15:31:23Z-2023‐09‐20T15:31:23Z", id="ISO timestamp range"),
+    ],
+)
+def test_update_time_range_invalid_time(time_range: str):
+    """
+    arrange: given an invalid time ranges.
+    act: when UpdateTimeRange class is instantiated through from_string method.
+    assert: ValueError is raised.
+    """
+    with pytest.raises(ValueError):
+        state.UpdateTimeRange.from_str(time_range)
+
+
+@pytest.mark.parametrize(
+    "time_range",
+    [
+        pytest.param("", id="empty string"),
+        pytest.param(None, id="None"),
+    ],
+)
+def test_update_time_range_empty_str(time_range):
+    """
+    arrange: given an empty time range config value.
+    act: when UpdateTimeRange class is instantiated through from_string method.
+    assert: None is returned.
+    """
+    assert (
+        state.UpdateTimeRange.from_str(time_range) is None
+    ), "Time range should not be instantiated."
+
+
+@pytest.mark.parametrize(
+    "time_range, expected_range",
+    [
+        pytest.param("0-3", (0, 3), id="valid time range single digits"),
+        pytest.param("00-03", (0, 3), id="valid time range single digits"),
+        pytest.param("21-3", (21, 3), id="overnight"),
+        pytest.param("21-03", (21, 3), id="overnight double digit"),
+    ],
+)
+def test_update_time_range_valid_time(time_range: str, expected_range: tuple[int, int]):
+    """
+    arrange: given a valid time range.
+    act: when UpdateTimeRange class is instantiated through from_string method.
+    assert: no exceptions are raised.
+    """
+    update_time_range = state.UpdateTimeRange.from_str(time_range)
+    assert update_time_range, "Expected time range to not be None."
+
+    assert update_time_range.start == expected_range[0], "Unexpected start time."
+    assert update_time_range.end == expected_range[1], "Unexpected end time."
+
+
+@pytest.mark.parametrize(
+    "patch_hour,time_range,expected_result",
+    [
+        pytest.param(3, "3-5", True, id="start hour"),
```

Comment:
What about just before?

Reply:
And after a past midnight should be added too
`pytest.param(2, "23-1", False, id="out of overnight range"),`

Reply:
Updated the test cases to include the before/edge/during/after cases :D Thanks for the review!